### PR TITLE
[one-cmds] Introduce target option

### PR DESCRIPTION
**Pull request by mhs4670go**
This commit introduces target option.

Related: #13161 
ONE-DCO-1.0-Signed-off-by: seongwoo <mhs4670go@naver.com>

### CODE REVIEW

File: compiler/one-cmds/one-profile
Comment (mhs4670go) [inline compiler/one-cmds/one-profile:1]: The change of this file is almost same with the one of `one-codegen`.

File: compiler/one-cmds/one-codegen
Comment (mhs4670go) [reply compiler/one-cmds/one-codegen:61]: I think the order looks good. Furthermore, changing the api right away looks fine because it is a private function of this `one-codegen` and it is used at the only one place.

File: compiler/one-cmds/one-profile
Comment (seanshpark) [inline compiler/one-cmds/one-profile:286]: maybe exceeds 100 col? if so, can you split?

File: compiler/one-cmds/onecc.template.cfg
Comment (seanshpark) [inline compiler/one-cmds/onecc.template.cfg:5]: below example uses `[backend]` but this line has `s`

File: compiler/one-cmds/tests/one-codegen_007.test
Comment (seanshpark) [inline compiler/one-cmds/tests/one-codegen_007.test:45]: cfg has `sunflower` but this checks `rose` .. is this negative test?

File: compiler/one-cmds/onecc.template.cfg
Comment (mhs4670go) [reply compiler/one-cmds/onecc.template.cfg:5]: Oh, my bad. I'll change it to `backend`.

File: compiler/one-cmds/tests/one-codegen_007.test
Comment (mhs4670go) [reply compiler/one-cmds/tests/one-codegen_007.test:45]: @seanshpark No. this test checks if the target option can be overwritten by the command line. The command line gives `rose` as a target option.

```bash
one-codegen -C ${configfile} -T rose > ${filename}.log 2>&1
```

**Review by seanshpark — APPROVED** [review]
LGTM

**Comment by seanshpark** [comment]
OK to land ? or need more reviews?

**Comment by mhs4670go** [comment]
> OK to land ? or need more reviews?

It's okay to land :)

**Comment by shs-park** [comment]
Although I am currently reviewing it, I will leave a comment if necessary even after it is landed.
😅

**Comment by seanshpark** [comment]
>  I will leave a comment if necessary even after it is landed.

I can wait for your review :)

File: compiler/one-cmds/tests/one_codegen_008.py
Comment (shs-park) [inline compiler/one-cmds/tests/one_codegen_008.py:11]: It seems that `DriverName`, `TargetOption` and `NormalOption` classes are not defined yet.
Are they going to be implemented later?

I couldn't find `target name` such as `rose` or `sunflower` in `one-codegen_008` test sequence.
Am I right..?

(Actually I couldn't fully understand the sequence of command schema. 😭)

**Review by shs-park — APPROVED** [review]
I left a question, PTAL.
Otherwise the code LGTM.
=)

File: compiler/one-cmds/tests/one_codegen_008.py
Comment (mhs4670go) [reply compiler/one-cmds/tests/one_codegen_008.py:11]: > Are they going to be implemented later?

I intentionally didn't implement the body. That classes just do the same with what enum does.

> I couldn't find target name such as rose or sunflower in one-codegen_008 test sequence.

The target is specified in the `one-codegen_008.cfg`;it says `target=one-codegen-008`. So, the test checks the output like below.
```bash
if grep -q "dummy-compile with one-codegen-008 target" "${outputfile}"; then
  echo "${filename_ext} SUCCESS"
  exit 0
fi
```

Comment (mhs4670go) [reply compiler/one-cmds/tests/one_codegen_008.py:11]: > Actually I couldn't fully understand the sequence of command schema. 

Please refer to the issue for the background on why the command schema was introduced. In a nutshell, it enables backend one-cmds tools such as _one-codegen_ or _one-profile_ to receive arguments in the configuration file according to the backends drivers' interface instead of the `command` key.

Anyway, I'm gonna have to write a document about that but let me elaborate a bit for now.

When the target input is given in the configuration file, _onecc_ would decide the actions based on these priorities.

[precondition] say, for instance, 'rose' target is given.

1. Check if the command schema exists according to the given `rose` target;the path should be like below.
```bash
 # `/usr/share/one/backends/command/${category}/${target}.py`
/usr/share/one/backends/command/codegen/rose.py
```
2. If the schema exists, _onecc_ could interpret the keys accordingly in the `one-codegen` section of the configuration file.
```ini
[backend]
target=rose
[one-codegen]
# $ dummy-compile --target rose -o sample.tvn sample.circle
o=sample.tvn
input=sample.circle
```
3. If the schema doesn't exist, _onecc_ would do as it normally does except one thing;it adds `--target rose` at the beginning of the command.
```ini
[backend]
target=rose
[one-codegen]
backend=dummy
# $ dummy-compile --target rose -o sample.tvn sample.circle
command=-o sample.tvn sample.circle
```

Comment (shs-park) [reply compiler/one-cmds/tests/one_codegen_008.py:11]: @mhs4670go,
Thanks for the detailed description 🙂

One more question, 
So the schema has higher priority than the `command` field in config?

Comment (mhs4670go) [reply compiler/one-cmds/tests/one_codegen_008.py:11]: @shs-park 

> So the schema has higher priority than the command field in config?

Yes. But, this is only when "target" is given.

**Pull request merged**
